include/xtensor-zarr/xzarr_aws_store.hpp

```diff
@@ -0,0 +1,286 @@
+        }
+        while (m_root.back() == '/')
+        {
+            m_root.pop_back();
+        }
+        std::size_t i = m_root.find('/');
+        if (i == std::string::npos)
+        {
+            m_bucket = m_root.c_str();
+            m_root = "";
+        }
+        else
+        {
+            m_bucket = m_root.substr(0, i).c_str();
+            m_root = m_root.substr(i + 1);
+        }
+    }
+
+    xzarr_aws_stream xzarr_aws_store::operator[](const std::string& key)
+    {
+        return xzarr_aws_stream((m_root + '/' + key).c_str(), m_bucket, m_client);
+    }
+
+    void xzarr_aws_store::set(const std::string& key, const std::vector<char>& value)
+    {
+        xzarr_aws_stream((m_root + '/' + key).c_str(), m_bucket, m_client) = value;
+    }
+
+    void xzarr_aws_store::set(const std::string& key, const std::string& value)
+    {
+        xzarr_aws_stream((m_root + '/' + key).c_str(), m_bucket, m_client) = value;
+    }
+
+    std::string xzarr_aws_store::get(const std::string& key)
+    {
+        return std::move(xzarr_aws_stream((m_root + '/' + key).c_str(), m_bucket, m_client));
+    }
+
+    void xzarr_aws_store::list_dir(const std::string& prefix, std::vector<std::string>& keys, std::vector<std::string>& prefixes)
+    {
+        Aws::S3::Model::ListObjectsRequest request;
+        std::string full_prefix = prefix;
+        if (!m_root.empty())
+        {
+            full_prefix = m_root + '/' + prefix;
+        }
+        request.WithBucket(m_bucket).WithPrefix(full_prefix.c_str());
+        auto outcome = m_client.ListObjects(request);
+        if (!outcome.IsSuccess())
+        {
+            auto err = outcome.GetError();
+            XTENSOR_THROW(std::runtime_error, std::string("Error: ListObjects: ") + err.GetExceptionName().c_str() + ": " + err.GetMessage().c_str());
+        }
+        Aws::Vector<Aws::S3::Model::Object> objects = outcome.GetResult().GetContents();
+
+        for (Aws::S3::Model::Object& object: objects)
+        {
+            auto key = object.GetKey();
+            std::size_t i = key.find('/');
+            if (i == std::string::npos)
+            {
+                keys.push_back(key.c_str());
+            }
+            else
+            {
+                key = key.substr(0, i + 1);
+                if (prefixes.size() == 0)
+                {
+                    prefixes.push_back(key.c_str());
+                }
+                else
+                {
+                    if (prefixes[prefixes.size() - 1] != key.c_str())
+                    {
+                        prefixes.push_back(key.c_str());
+                    }
+                }
+            }
+        }
+    }
+
+    std::vector<std::string> xzarr_aws_store::list()
+    {
+        return list_prefix("");
+    }
+
+    std::vector<std::string> xzarr_aws_store::list_prefix(const std::string& prefix)
+    {
+        std::vector<std::string> keys;
+        Aws::S3::Model::ListObjectsRequest request;
```

Comment:
It is better to declare variables just before using them instead of declaring them at the beginning of the function. Most of the time, this allows you to declare and initialize them on the same line. Therefore, these two lines can be removed.